compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotter.kt

```diff
@@ -108,114 +179,57 @@ object ClassSnapshotter {
 
-        fun BasicClassInfo.isTransitivelyInaccessible(): Boolean {
-            classIsInaccessible[this]?.let { return it }
+    /** The directory or jar to read. */
+    private val root: Path
 
-            val inaccessible = if (isInaccessible()) {
-                true
-            } else classId.outerClassId?.let { outerClassId ->
-                classIdToClassInfo[outerClassId]?.isTransitivelyInaccessible()
-                // If we can't find the outer class from the given list of classes (which could happen with faulty jars), we assume that
-                // the class is accessible (see this method's kdoc).
-                    ?: false
-            } ?: false
+    private val closer = Closer.create()
 
-            return inaccessible.also {
-                classIsInaccessible[this] = inaccessible
-            }
+    init {
+        root = if (directoryOrJar.isDirectory) {
+            directoryOrJar.toPath()
+        } else {
+            check(directoryOrJar.isFile && directoryOrJar.path.endsWith(".jar", ignoreCase = true))
+            FileSystems.newFileSystem(directoryOrJar.toPath(), null as ClassLoader?)
+                .also { closer.register(it) }
+                .rootDirectories.single()
         }
-
-        return classesInfo.filterTo(mutableSetOf()) { it.isTransitivelyInaccessible() }
     }
-}
-
-/** Utility to read the contents of a directory or jar. */
-private object DirectoryOrJarContentsReader {
 
     /**
-     * Returns a map from Unix-style relative paths of entries to their contents. The paths are relative to the given container (directory
-     * or jar).
-     *
-     * The map entries need to satisfy the given filter.
+     * Returns the Unix-style relative paths of all entries under the containing directory or jar which satisfy the given [filter].
      *
-     * The map entries are sorted based on their Unix-style relative paths (to ensure deterministic results across filesystems).
+     * The paths are in Unix style and are sorted to ensure deterministic results across platforms.
      *
-     * Note: If a jar has duplicate entries after filtering, only the first one is retained.
+     * If a jar has duplicate entries, only unique relative paths are kept in the returned list.
      */
-    fun read(
-        directoryOrJar: File,
-        entryFilter: ((unixStyleRelativePath: String, isDirectory: Boolean) -> Boolean)? = null
-    ): LinkedHashMap<String, ByteArray> {
-        return if (directoryOrJar.isDirectory) {
-            readDirectory(directoryOrJar, entryFilter)
-        } else {
-            check(directoryOrJar.isFile && directoryOrJar.path.endsWith(".jar", ignoreCase = true))
-            readJar(directoryOrJar, entryFilter)
-        }
+    fun getUnixStyleRelativePaths(filter: (unixStyleRelativePath: String, isDirectory: Boolean) -> Boolean): List<String> {
+        return Files.walk(root).use {
+            // Use `Set` to de-duplicate entries
```

Comment:
Can we have duplicate entries in a jar? If we do have them, can compiler load 2 classes e.g. one with FQCN `A.B` and another `A$B`?

Reply:
I believe only faulty jars will have duplicate entries.

For valid jars, it's not possible to have `ClassId` `A.B` and `A$B` in the same jar as they have the same file name `A$B.class` (the compiler will also throw an error if we try to define those 2 classes in the same module).

If those two classes are located in two different jars on the classpath, the compiler will select the first one.

So I guess in the same way it should be safe to de-duplicate entries here?

Reply:
Sounds good, I agree it should be safe to do this.